File: Java/VetoGame.java
import java.util.LinkedList;
import java.util.Queue;

public class VetoGame {
    //
    private final String[] preferences;
    
    public VetoGame(String[] preferences) {
        this.preferences = preferences;
    }
    
    public String[] getPreferences() {
        return preferences;
    }
    
    //
    private VetoGameState searchFromRoot() {
        int n = preferences.length;
        
        if (preferences[0].length() != n + 1) {
            throw new IllegalArgumentException(
                "`preferences` must have `n` columns (players) and `n + 1` rows (options)."
            );
        }
        
        VetoGameState root = new VetoGameState(preferences[0], '\0', 0, null);
        Queue<VetoGameState> nodeQueue = new LinkedList<>();
        nodeQueue.add(root);
        
        while (nodeQueue.peek().getDepth() < n) {
            VetoGameState node = nodeQueue.remove();
            String state = node.getState();
            int stateSize = state.length();
            
            for (int i = 0; i < stateSize - 1; i++) {
                String stateChild = state.substring(0, i);
                stateChild += state.substring(i + 1);
                
                VetoGameState child = new VetoGameState(
                    stateChild,
                    state.charAt(i),
                    node.getDepth() + 1,
                    node
                );
                
                node.addChild(child);
                nodeQueue.add(child);
            }
            
            VetoGameState child = new VetoGameState(
                state.substring(0, stateSize - 1),
                state.charAt(stateSize - 1),
                node.getDepth() + 1,
                node
            );
            
            node.addChild(child);
            nodeQueue.add(child);
        }
        
        while (nodeQueue.peek().getDepth() == n) {
            VetoGameState node = nodeQueue.remove();
            node.setOutcome(node.getState().charAt(0));
            nodeQueue.add(node.getParent());
        }
        
        while (nodeQueue.peek().getDepth() > 0) {
            VetoGameState node = nodeQueue.peek();
            VetoGameState parent = node.getParent();
            
            while (!nodeQueue.isEmpty() && nodeQueue.peek().getParent() == parent) {
                node = nodeQueue.remove();
                node.findOutcome(preferences);
            }
            
            nodeQueue.add(node.getParent());
        }
        
        root.findOutcome(preferences);
        return root;
    }
    
    public VetoGameResult findSolution() {
        VetoGameState node = searchFromRoot();
        String solution = "";
        
        while (node.getDepth() < preferences.length) {
            VetoGameState child = node.getChildren().get(0);
            solution += child.getDeletion();
            node = child;
        }
        
        return new VetoGameResult(solution, node.getOutcome());
    }
}

File: Julia/VetoGame.jl
module VetoGame
    # EXPORTS AND IMPORTS
    export VetoGameResult, solve_veto_game
    using Combinatorics: combinations
    using DataStructures: Queue, dequeue!, enqueue!
    
    # STRUCT: `VetoGameResult`
    struct VetoGameResult{T}
        solution::Vector{T}
        outcome::T
    end
    
    function Base.getproperty(result::VetoGameResult, prop::Symbol)
        if prop in fieldnames(VetoGameResult)
            out = getfield(result, prop)
        elseif prop == :to_string
            rounds = Vector{String}(undef, length(result.solution) + 1)
            rounds[1:(end - 1)] .= [
                "Player $i vetoes $option\n" for (i, option) in enumerate(result.solution)
            ]
            rounds[end] = "\nOutcome: $(result.outcome)"
            out = join(rounds)
        else
            throw(ArgumentError("type VetoGameResult has no field $prop"))
        end
        
        return out
    end
    
    function Base.propertynames(::VetoGameResult)
        return (:solution, :outcome, :to_string)
    end
    
    # MUTABLE STRUCT: `VetoGameState`
    mutable struct VetoGameState{T}
        state::Vector{T}
        parent::Union{Nothing, VetoGameState{T}}
        children::Vector{VetoGameState{T}}
        outcome::Union{Missing, T}
        depth::Int64
    end
    
    # FUNCTION: `solve_veto_game`
    function solve_veto_game(preferences::Matrix{T}) where T
        (m, n) = size(preferences)
        
        if m != n + 1
            throw(ArgumentError(
                "`preferences` must have `n` columns (players) and `n + 1` rows (options)."
            ))
        end
        
        root = VetoGameState{T}(unique(preferences), nothing, VetoGameState{T}[], missing, 0)
        Q = Queue{VetoGameState{T}}()
        enqueue!(Q, root)
        
        while first(Q).depth < n
            node = dequeue!(Q)
            depth = node.depth
            
            for comb in combinations(node.state, n - depth)
                child = VetoGameState{T}(comb, node, VetoGameState{T}[], missing, depth + 1)
                push!(node.children, child)
                enqueue!(Q, child)
            end
        end
        
        while first(Q).depth == n
            node = dequeue!(Q)
            node.outcome = node.state[1]
            enqueue!(Q, node.parent)
        end
        
        while first(Q).depth > 0
            parent = first(Q).parent
            
            while !isempty(Q) && first(Q).parent == parent
                node = dequeue!(Q)
                set_outcome!(node, preferences)
            end
            
            enqueue!(Q, node.parent)
        end
        
        set_outcome!(root, preferences)
        solution = Vector{T}(undef, n)
        node = root
        
        while node.depth < n
            child = node.children[1]
            solution[node.depth + 1] = setdiff(node.state, child.state)[1]
            node = child
        end
        
        return VetoGameResult{T}(solution, node.outcome)
    end
    
    # FUNCTION: `set_outcome!`
    function set_outcome!(node::VetoGameState{T}, preferences::Matrix{T}) where T
        player_prefs = @view preferences[:, node.depth + 1]
        sort!(node.children, by = x -> findfirst(player_prefs .== x.outcome))
        
        child = node.children[1]
        node.children = [child]
        node.outcome = child.outcome
        return node
    end
end